{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Min position: 30798252\n",
      "Max position: 31568419\n"
     ]
    }
   ],
   "source": [
    "def find_min_max_positions_bim(bim_file):\n",
    "    with open(bim_file, 'r') as f:\n",
    "        positions = [int(line.split()[3]) for line in f]\n",
    "    \n",
    "    min_pos = min(positions)\n",
    "    max_pos = max(positions)\n",
    "    \n",
    "    return min_pos, max_pos\n",
    "\n",
    "bim_file = \"data/fourier_ls-chr6-1167_train.bim\"\n",
    "idx1, idx2 = find_min_max_positions_bim(bim_file)\n",
    "\n",
    "print(f\"Min position: {idx1}\")\n",
    "print(f\"Max position: {idx2}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[XCFtools] Convert from/to XCF files\n",
      "  * Authors       : Olivier DELANEAU, University of Lausanne\n",
      "  * Contact       : olivier.delaneau@gmail.com\n",
      "  * Version       : 1.0.0 / commit = 48abb98 / release = 2023-07-25\n",
      "  * Run date      : 23/10/2024 - 10:45:47\n",
      "\n",
      "Files:\n",
      "  * Input BCF     : [fourier_ls-chr6-1167_train.bcf]\n",
      "  * Output XCF    : [train_xcf.bcf]\n",
      "\n",
      "Parameters:\n",
      "  * Seed    : 15052011\n",
      "  * Threads : 8 threads\n",
      "  * MAF     : 0.03125\n",
      "\n",
      "Converting from BCF to XCF [Sparse/Haplotype]\n",
      "  * Region        : 6\n",
      "  * Contig        : 6\n",
      "  * Min MAF       : 0.03125\n",
      "  * #samples = 225245\n",
      "  * Number of BCF records processed: Nc=0/ Nr=0\n",
      "\n",
      "Finalization:\n",
      "  * Total running time = 0 seconds\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "import os\n",
    "train_bcf = 'data/fourier_ls-chr6-1167_train.bcf'\n",
    "\n",
    "result = subprocess.run([\n",
    "        '/scratch2/prateek/impute5_v1.2.0/xcftools_static', 'view',\n",
    "        '-i', train_bcf,\n",
    "        '-o', f\"train_xcf.bcf\",\n",
    "        '-O', 'sh',\n",
    "        '-r', '6',\n",
    "        '-T8',\n",
    "        '-m', '0.03125'\n",
    "    ])\n",
    "\n",
    "train_xcf = \"train_xcf.bcf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def process_plink_data(plink_file_prefix):\n",
    "    vcf_file = f\"data/{plink_file_prefix}.vcf\"\n",
    "    # print(\"Exporting PLINK data to VCF...\")\n",
    "    export_command = [\n",
    "        \"../plink2\", \"--bfile\", plink_file_prefix,\n",
    "        \"--recode\", \"vcf\", \"--out\", plink_file_prefix,\n",
    "        \"--silent\"\n",
    "    ]\n",
    "    \n",
    "    if not os.path.exists(vcf_file):\n",
    "        subprocess.run(export_command, check=True)\n",
    "\n",
    "    bcftools_command = [\n",
    "        \"../bcftools/bcftools\", \"view\", \"-Oz\", \"-o\", f\"data/{plink_file_prefix}.bcf\", vcf_file\n",
    "    ]\n",
    "\n",
    "    if not os.path.exists(f\"data/{plink_file_prefix}.bcf\"):\n",
    "        subprocess.run(bcftools_command, check=True)\n",
    "\n",
    "        # print(\"BCF conversion complete.\")\n",
    "\n",
    "        # print(\"Adding AC field..\")\n",
    "        subprocess.run([\n",
    "        '../bcftools/bcftools', '+fill-tags', f'data/{plink_file_prefix}.bcf', '-Ob', '-o', f'data/{plink_file_prefix}_AC.bcf', '--', '-t', 'AN,AC'\n",
    "        ], stdout=subprocess.DEVNULL)\n",
    "        # print(\"Adding index file..\")\n",
    "        subprocess.run(['../bcftools/bcftools', 'index', f'data/{plink_file_prefix}_AC.bcf'])\n",
    "\n",
    "    if not os.path.exists(f\"{plink_file_prefix}_xcf.bcf\"):\n",
    "        subprocess.run([\n",
    "            '../impute5_v1.2.0/xcftools_static', 'view',\n",
    "            '-i', f\"data/{plink_file_prefix}_AC.bcf\",\n",
    "            '-o', f\"data/{plink_file_prefix}_xcf.bcf\",\n",
    "            '-O', 'sh',\n",
    "            '-r', '6',\n",
    "            '-T8',\n",
    "            '-m', '0.03125'\n",
    "        ])\n",
    "\n",
    "def process_plink_data_with_drop(plink_file_prefix, rs_ids):\n",
    "    vcf_file = f\"data/{plink_file_prefix}.vcf\"\n",
    "    # print(\"Exporting PLINK data to VCF...\")\n",
    "    export_command = [\n",
    "        \"../plink2\", \"--bfile\", f\"data/{plink_file_prefix}\",\n",
    "        \"--recode\", \"vcf\", \"--out\", f\"data/{plink_file_prefix}\",\n",
    "        \"--silent\"\n",
    "    ]\n",
    "    subprocess.run(export_command)\n",
    "\n",
    "    modified_vcf_file = f\"modified_test.vcf\"\n",
    "    with open(vcf_file, 'r') as f:\n",
    "        lines = f.readlines()\n",
    "\n",
    "    with open(modified_vcf_file, 'w') as out_file:\n",
    "        for line in lines:\n",
    "            if line.startswith('#'):\n",
    "                out_file.write(line)\n",
    "                continue\n",
    "            \n",
    "            fields = line.strip().split('\\t')\n",
    "            snp_id = fields[2]\n",
    "\n",
    "            if snp_id in rs_ids:\n",
    "                # fields[9:] = ['./.' for _ in fields[9:]]\n",
    "                continue\n",
    "                \n",
    "            out_file.write('\\t'.join(fields) + '\\n')\n",
    "\n",
    "    # print(\"VCF modification complete.\")\n",
    "\n",
    "    bcftools_command = [\n",
    "        \"../bcftools/bcftools\", \"view\", \"-Oz\", '--threads', '32', \"-o\", \"modified_test.bcf\", modified_vcf_file\n",
    "    ]\n",
    "    subprocess.run(bcftools_command)\n",
    "\n",
    "    # print(\"BCF conversion complete.\")\n",
    "\n",
    "    # print(\"Adding AC field..\")\n",
    "    subprocess.run([\n",
    "     '../bcftools/bcftools', '+fill-tags', 'modified_test.bcf', '-Ob', '-o', 'modified_test_AC.bcf', '--', '-t', 'AN,AC'\n",
    "    ], stdout=subprocess.DEVNULL)\n",
    "    # print(\"Adding index file..\")\n",
    "    subprocess.run(['../bcftools/bcftools', 'index', 'modified_test_AC.bcf'])\n",
    "\n",
    "    # print('Converting to XCF...')\n",
    "    subprocess.run([\n",
    "        '../impute5_v1.2.0/xcftools_static', 'view',\n",
    "        '-i', 'modified_test_AC.bcf',\n",
    "        '-o', \"modified_test_xcf.bcf\",\n",
    "        '-O', 'sh',\n",
    "        '-r', '6',\n",
    "        '-T8',\n",
    "        '-m', '0.03125'\n",
    "    ])\n",
    "\n",
    "def extract_imputed_genotype_array(vcf_file, target_snp):\n",
    "\n",
    "    with open(vcf_file, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "        \n",
    "        for line in lines:\n",
    "            if line.startswith('#'):\n",
    "                continue  # Skip header lines\n",
    "            \n",
    "            fields = line.strip().split('\\t')\n",
    "            snp_id = f\"{fields[0]}:{fields[1]}\"  # Format: chr:pos\n",
    "            \n",
    "            if snp_id == target_snp:\n",
    "                genotype_data = fields[9:]  # Genotype fields for all samples\n",
    "                genotype_array = []\n",
    "\n",
    "                # Parse genotype data\n",
    "                for genotype in genotype_data:\n",
    "                    gt_info = genotype.split(':')[0]  # Take the genotype part (GT)\n",
    "                    # Convert genotype format to 0, 1, or 2\n",
    "                    if gt_info == \"0|0\":\n",
    "                        genotype_array.append(0)\n",
    "                    elif gt_info == \"0|1\" or gt_info == \"1/0\":\n",
    "                        genotype_array.append(1)\n",
    "                    elif gt_info == \"1|1\":\n",
    "                        genotype_array.append(2)\n",
    "                    else:\n",
    "                        genotype_array.append(-1)  # for missing or unknown genotypes\n",
    "                \n",
    "                return genotype_array\n",
    "    \n",
    "    return None  # SNP not found in the VCF file\n",
    "\n",
    "\n",
    "def extract_test_genotype_array(vcf_file, target_snp):\n",
    "\n",
    "    with open(vcf_file, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "        \n",
    "        for line in lines:\n",
    "            if line.startswith('#'):\n",
    "                continue  # Skip header lines\n",
    "            \n",
    "            fields = line.strip().split('\\t')\n",
    "            snp_id = f\"{fields[0]}:{fields[1]}\"  # Format: chr:pos\n",
    "            \n",
    "            if snp_id == target_snp:\n",
    "                genotype_data = fields[9:]  # Genotype fields for all samples\n",
    "                genotype_array = []\n",
    "\n",
    "                # Parse genotype data\n",
    "                for genotype in genotype_data:\n",
    "                    # Convert genotype format to 0, 1, or 2\n",
    "                    if genotype == \"0/0\":\n",
    "                        genotype_array.append(0)\n",
    "                    elif genotype == \"0/1\" or genotype == \"1/0\":\n",
    "                        genotype_array.append(1)\n",
    "                    elif genotype == \"1/1\":\n",
    "                        genotype_array.append(2)\n",
    "                    else:\n",
    "                        genotype_array.append(-1)  # for missing or unknown genotypes\n",
    "                \n",
    "                return genotype_array\n",
    "    \n",
    "    return None  # SNP not found in the VCF file\n",
    "\n",
    "def compute_r_squared(array1, array2):\n",
    "    # Ensure both arrays are NumPy arrays\n",
    "    array1 = np.array(array1)\n",
    "    array2 = np.array(array2)\n",
    "\n",
    "    # Calculate the correlation coefficient\n",
    "    correlation_matrix = np.corrcoef(array1, array2)\n",
    "    correlation_xy = correlation_matrix[0, 1]\n",
    "    \n",
    "    # Calculate R^2\n",
    "    r_squared = correlation_xy ** 2\n",
    "    \n",
    "    return r_squared"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dropping SNP: 6:30798252\n",
      "PLINK v2.0.0-a.6LM AVX2 AMD (20 Oct 2024)          cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to fourier_ls-chr6-1167_test.log.\n",
      "Options in effect:\n",
      "  --bfile fourier_ls-chr6-1167_test\n",
      "  --export vcf\n",
      "  --out fourier_ls-chr6-1167_test\n",
      "\n",
      "Start time: Wed Oct 23 15:29:11 2024\n",
      "128674 MiB RAM detected, ~113002 available; reserving 64337 MiB for main\n",
      "workspace.\n",
      "Using up to 32 threads (change this with --threads).\n",
      "56312 samples (30180 females, 26132 males; 56312 founders) loaded from\n",
      "fourier_ls-chr6-1167_test.fam.\n",
      "1167 variants loaded from fourier_ls-chr6-1167_test.bim.\n",
      "Note: No phenotype data present.\n",
      "--export vcf to fourier_ls-chr6-1167_test.vcf ... 1010111112121313141415151616171718181919202021212222232324242525262627272828292930303131323233333434353536363737383839394040414142424343444445454646474748484949505051515252535354545555565657575858595960606161626263636464656566666767686869697070717172727373747475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "End time: Wed Oct 23 15:29:13 2024\n",
      "\n",
      "[XCFtools] Convert from/to XCF files\n",
      "  * Authors       : Olivier DELANEAU, University of Lausanne\n",
      "  * Contact       : olivier.delaneau@gmail.com\n",
      "  * Version       : 1.0.0 / commit = 48abb98 / release = 2023-07-25\n",
      "  * Run date      : 23/10/2024 - 15:29:52\n",
      "\n",
      "Files:\n",
      "  * Input BCF     : [modified_test_AC.bcf]\n",
      "  * Output XCF    : [modified_test_xcf.bcf]\n",
      "\n",
      "Parameters:\n",
      "  * Seed    : 15052011\n",
      "  * Threads : 8 threads\n",
      "  * MAF     : 0.03125\n",
      "\n",
      "Converting from BCF to XCF [Sparse/Haplotype]\n",
      "  * Region        : 6\n",
      "  * Contig        : 6\n",
      "  * Min MAF       : 0.03125\n",
      "  * #samples = 56312\n",
      "  * Number of BCF records processed: Nc=898/ Nr=268\n",
      "\n",
      "Finalization:\n",
      "  * Total running time = 2 seconds\n",
      "\n",
      "[IMPUTE5] Imputation of phased SNP array data from large reference panels\n",
      "  * Authors             : Simone Rubinacci & Jonathan Marchini, 2023\n",
      "  * Contact             : rubinacci.simone@gmail.com\n",
      "  * Version       \t: IMPUTE5 v1.2.0 / commit = 6e3d41e / release = 2023-08-21\n",
      "  * Citation\t        : PLOS Genetics 16(11): e1009049 (2020). DOI: https://doi.org/10.1371/journal.pgen.1009049\n",
      "  * Licence\t        : IMPUTE5 is freely available only for academic use. To see rules for non-academic use, please read the LICENCE.\n",
      "  * Run date            : 23/10/2024 - 15:29:55\n",
      "\n",
      "Files:\n",
      "  * Input phased array  : [modified_test_xcf.bcf]\n",
      "  * Reference panel     : [fourier_ls-chr6-1167_train_xcf.bcf]\n",
      "  * Output file         : [imputed_0.vcf]\n",
      "  * Output format       : [VCF format | NO compression | with CSI index]\n",
      "\n",
      "\u001b[33mWARNING: \u001b[0mPlain VCF file format with no compression detected as output: this format is strongly discouraged as does not allow the creation of a CSI index, can create very large files. Please use a compressed file format if possible (vcf.gz, bcf).\n",
      "  * Output LOG          : [imputed_0.log]\n",
      "\n",
      "IMPUTE5 parameters:\n",
      "  * Imputation model    : [Reference panel imputation]\n",
      "  * Input region        : [6:30798252-31568419]\n",
      "  * Output region       : [6:30798252-30798253]\n",
      "  * Sparse MAF          : [0.03125]\n",
      "  * Recombination rates : [Constant rate of 1cM/Mb]\n",
      "  * Ploidy              : [Only diploid samples in region]\n",
      "\n",
      "Model parameters:\n",
      "  * Ne [eff. pop. size] : [1000000]\n",
      "  * Imputation err. rate: [0.0001]\n",
      "\n",
      "Selection parameters:\n",
      "  * K pbwt              : [1500]\n",
      "  * PBWT max depth      : [10]\n",
      "  * PBWT min depth      : [2]\n",
      "  * PBWT modulo (cM)    : [0.02]\n",
      "\n",
      "Test statistics parameters:\n",
      "  * Surfbat test        : [NO]\n",
      "  * Surfbat MAF         : [0.01]\n",
      "  * Surfbat INFO        : [0.3]\n",
      "\n",
      "Output\n",
      "  * FORMAT/GT           : [YES]\n",
      "  * FORMAT/DS           : [YES]\n",
      "  * FORMAT/GP           : [YES]\n",
      "  * FORMAT/AP           : [NO]\n",
      "  * FORMAT/SAP          : [NO]\n",
      "  * Buffer              : [NO]\n",
      "  * CSI index           : [YES]\n",
      "\n",
      "Other parameters\n",
      "  * Seed                : [42]\n",
      "  * #Threads            : [32]\n",
      "\n",
      "Initialisation:\n",
      "  * XCF scanning\t[done]\t\t[Li=1167 | Lg= 1166 | Lt=0]\t(0.26s)\n",
      "  * XCF parsing\t\t[done]\t\t[Li=1167 | Lg= 1166 | Lt=0]\t(1.28s)\n",
      "  * Common transpose\t[ref]\t\t[var2hap]\t\t\t(0.55s)\n",
      "  * Rare transpose\t[ref]\t\t[var2hap]\t[nrar=0.0231814]\t(0.00s)\n",
      "\n",
      "\u001b[33mWARNING: \u001b[0mcM constant\t[1cM=1Mb]\t\t\t\t\t(0.00s)\n",
      "\n",
      "Samples statistics\n",
      "  * Reference panel \t[Nrh=450490]\t[L=1167]\t[Lb=1166]\n",
      "    \tImputation reg.\t[Lg=0]\t[Lr=1]\t[Lc=1166]\n",
      "  * Target panel \t[Nth=112624]\t[L=1166]\t[Lb=1166]\n",
      "    \tImputation reg.\t[Lg=0]\t[Lt=0]\n",
      "\n",
      "Region statistics\n",
      "  * Buffer region \t[6:30798252-31568419]\n",
      "  * Imputation region\t[6:30798252-30798253]\n",
      "  * Region span   \t[0.77 Mb]\t[0.77 cM]\n",
      "\n",
      "PBWT state selection\n",
      "  * PBWT stats\t\t[Div sel]\t[n_stored=36]\t[done]\t\t(11.99s)\n",
      "  * Geno transpose\t[ref]\t\t[var2hap]\t\t\t(0.54s)\n",
      "\n",
      "Imputing variants\n",
      "  * HMM imputation\t[done]\t[#states=143.7 / mono=0.0% / %srd=0]\t(4.82s)\n",
      "\n",
      "Finalization\n",
      "  * VCF writing \t[done]\t\t[RAW  / N=56312 / L=1167]\t(0.03s)\n",
      "\n",
      "\u001b[33mWARNING: \u001b[0mPlain VCF file format with no compression as output: no CSI index can be created. Uncompressed VCF file format is strongly discouraged. Please use vcf.gz or bcf file format.\n",
      "\n",
      "  * Output file         : [imputed_0.vcf]\n",
      "  * Output format       : [VCF format | NO compression | with CSI index]\n",
      "\n",
      "\u001b[33mWARNING: \u001b[0mPlain VCF file format with no compression detected as output: this format is strongly discouraged as does not allow the creation of a CSI index, can create very large files. Please use a compressed file format if possible (vcf.gz, bcf).\n",
      "\n",
      "Total running time [19 seconds] (19s)\n",
      "0.4888814999377982\n",
      "File modified_test_xcf.vcf not found for deletion\n",
      "Dropping SNP: 6:30800326\n",
      "PLINK v2.0.0-a.6LM AVX2 AMD (20 Oct 2024)          cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to fourier_ls-chr6-1167_test.log.\n",
      "Options in effect:\n",
      "  --bfile fourier_ls-chr6-1167_test\n",
      "  --export vcf\n",
      "  --out fourier_ls-chr6-1167_test\n",
      "\n",
      "Start time: Wed Oct 23 15:30:15 2024\n",
      "128674 MiB RAM detected, ~112846 available; reserving 64337 MiB for main\n",
      "workspace.\n",
      "Using up to 32 threads (change this with --threads).\n",
      "56312 samples (30180 females, 26132 males; 56312 founders) loaded from\n",
      "fourier_ls-chr6-1167_test.fam.\n",
      "1167 variants loaded from fourier_ls-chr6-1167_test.bim.\n",
      "Note: No phenotype data present.\n",
      "--export vcf to fourier_ls-chr6-1167_test.vcf ... 1010111112121313141415151616171718181919202021212222232324242525262627272828292930303131323233333434353536363737383839394040414142424343444445454646474748484949505051515252535354545555565657575858595960606161626263636464656566666767686869697070717172727373747475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "End time: Wed Oct 23 15:30:18 2024\n",
      "\n",
      "[XCFtools] Convert from/to XCF files\n",
      "  * Authors       : Olivier DELANEAU, University of Lausanne\n",
      "  * Contact       : olivier.delaneau@gmail.com\n",
      "  * Version       : 1.0.0 / commit = 48abb98 / release = 2023-07-25\n",
      "  * Run date      : 23/10/2024 - 15:30:54\n",
      "\n",
      "Files:\n",
      "  * Input BCF     : [modified_test_AC.bcf]\n",
      "  * Output XCF    : [modified_test_xcf.bcf]\n",
      "\n",
      "Parameters:\n",
      "  * Seed    : 15052011\n",
      "  * Threads : 8 threads\n",
      "  * MAF     : 0.03125\n",
      "\n",
      "Converting from BCF to XCF [Sparse/Haplotype]\n",
      "  * Region        : 6\n",
      "  * Contig        : 6\n",
      "  * Min MAF       : 0.03125\n",
      "  * #samples = 56312\n",
      "  * Number of BCF records processed: Nc=897/ Nr=269\n",
      "\n",
      "Finalization:\n",
      "  * Total running time = 2 seconds\n",
      "\n",
      "[IMPUTE5] Imputation of phased SNP array data from large reference panels\n",
      "  * Authors             : Simone Rubinacci & Jonathan Marchini, 2023\n",
      "  * Contact             : rubinacci.simone@gmail.com\n",
      "  * Version       \t: IMPUTE5 v1.2.0 / commit = 6e3d41e / release = 2023-08-21\n",
      "  * Citation\t        : PLOS Genetics 16(11): e1009049 (2020). DOI: https://doi.org/10.1371/journal.pgen.1009049\n",
      "  * Licence\t        : IMPUTE5 is freely available only for academic use. To see rules for non-academic use, please read the LICENCE.\n",
      "  * Run date            : 23/10/2024 - 15:30:57\n",
      "\n",
      "Files:\n",
      "  * Input phased array  : [modified_test_xcf.bcf]\n",
      "  * Reference panel     : [fourier_ls-chr6-1167_train_xcf.bcf]\n",
      "  * Output file         : [imputed_1.vcf]\n",
      "  * Output format       : [VCF format | NO compression | with CSI index]\n",
      "\n",
      "\u001b[33mWARNING: \u001b[0mPlain VCF file format with no compression detected as output: this format is strongly discouraged as does not allow the creation of a CSI index, can create very large files. Please use a compressed file format if possible (vcf.gz, bcf).\n",
      "  * Output LOG          : [imputed_1.log]\n",
      "\n",
      "IMPUTE5 parameters:\n",
      "  * Imputation model    : [Reference panel imputation]\n",
      "  * Input region        : [6:30798252-31568419]\n",
      "  * Output region       : [6:30800326-30800327]\n",
      "  * Sparse MAF          : [0.03125]\n",
      "  * Recombination rates : [Constant rate of 1cM/Mb]\n",
      "  * Ploidy              : [Only diploid samples in region]\n",
      "\n",
      "Model parameters:\n",
      "  * Ne [eff. pop. size] : [1000000]\n",
      "  * Imputation err. rate: [0.0001]\n",
      "\n",
      "Selection parameters:\n",
      "  * K pbwt              : [1500]\n",
      "  * PBWT max depth      : [10]\n",
      "  * PBWT min depth      : [2]\n",
      "  * PBWT modulo (cM)    : [0.02]\n",
      "\n",
      "Test statistics parameters:\n",
      "  * Surfbat test        : [NO]\n",
      "  * Surfbat MAF         : [0.01]\n",
      "  * Surfbat INFO        : [0.3]\n",
      "\n",
      "Output\n",
      "  * FORMAT/GT           : [YES]\n",
      "  * FORMAT/DS           : [YES]\n",
      "  * FORMAT/GP           : [YES]\n",
      "  * FORMAT/AP           : [NO]\n",
      "  * FORMAT/SAP          : [NO]\n",
      "  * Buffer              : [NO]\n",
      "  * CSI index           : [YES]\n",
      "\n",
      "Other parameters\n",
      "  * Seed                : [42]\n",
      "  * #Threads            : [32]\n",
      "\n",
      "Initialisation:\n",
      "  * XCF scanning\t[done]\t\t[Li=1167 | Lg= 1166 | Lt=0]\t(0.19s)\n",
      "  * XCF parsing\t\t[done]\t\t[Li=1167 | Lg= 1166 | Lt=0]\t(1.22s)\n",
      "  * Common transpose\t[ref]\t\t[var2hap]\t\t\t(0.40s)\n",
      "  * Rare transpose\t[ref]\t\t[var2hap]\t[nrar=0]\t(0.00s)\n",
      "\n",
      "\u001b[33mWARNING: \u001b[0mcM constant\t[1cM=1Mb]\t\t\t\t\t(0.00s)\n",
      "\n",
      "Samples statistics\n",
      "  * Reference panel \t[Nrh=450490]\t[L=1167]\t[Lb=1166]\n",
      "    \tImputation reg.\t[Lg=0]\t[Lr=0]\t[Lc=1167]\n",
      "  * Target panel \t[Nth=112624]\t[L=1166]\t[Lb=1166]\n",
      "    \tImputation reg.\t[Lg=0]\t[Lt=0]\n",
      "\n",
      "Region statistics\n",
      "  * Buffer region \t[6:30798252-31568419]\n",
      "  * Imputation region\t[6:30800326-30800327]\n",
      "  * Region span   \t[0.77 Mb]\t[0.77 cM]\n",
      "\n",
      "PBWT state selection\n",
      "  * PBWT stats\t\t[Div sel]\t[n_stored=36]\t[done]\t\t(12.37s)\n",
      "  * Geno transpose\t[ref]\t\t[var2hap]\t\t\t(0.60s)\n",
      "\n",
      "Imputing variants\n",
      "  * HMM imputation\t[done]\t[#states=140.2 / mono=0.0% / %srd=0]\t(4.99s)\n",
      "\n",
      "Finalization\n",
      "  * VCF writing \t[done]\t\t[RAW  / N=56312 / L=1167]\t(0.05s)\n",
      "\n",
      "\u001b[33mWARNING: \u001b[0mPlain VCF file format with no compression as output: no CSI index can be created. Uncompressed VCF file format is strongly discouraged. Please use vcf.gz or bcf file format.\n",
      "\n",
      "  * Output file         : [imputed_1.vcf]\n",
      "  * Output format       : [VCF format | NO compression | with CSI index]\n",
      "\n",
      "\u001b[33mWARNING: \u001b[0mPlain VCF file format with no compression detected as output: this format is strongly discouraged as does not allow the creation of a CSI index, can create very large files. Please use a compressed file format if possible (vcf.gz, bcf).\n",
      "\n",
      "Total running time [19 seconds] (19s)\n",
      "0.9737785565499684\n",
      "File modified_test_xcf.vcf not found for deletion\n",
      "Dropping SNP: 6:30800577\n",
      "PLINK v2.0.0-a.6LM AVX2 AMD (20 Oct 2024)          cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to fourier_ls-chr6-1167_test.log.\n",
      "Options in effect:\n",
      "  --bfile fourier_ls-chr6-1167_test\n",
      "  --export vcf\n",
      "  --out fourier_ls-chr6-1167_test\n",
      "\n",
      "Start time: Wed Oct 23 15:31:17 2024\n",
      "128674 MiB RAM detected, ~112847 available; reserving 64337 MiB for main\n",
      "workspace.\n",
      "Using up to 32 threads (change this with --threads).\n",
      "56312 samples (30180 females, 26132 males; 56312 founders) loaded from\n",
      "fourier_ls-chr6-1167_test.fam.\n",
      "1167 variants loaded from fourier_ls-chr6-1167_test.bim.\n",
      "Note: No phenotype data present.\n",
      "--export vcf to fourier_ls-chr6-1167_test.vcf ... 1010111112121313141415151616171718181919202021212222232324242525262627272828292930303131323233333434353536363737383839394040414142424343444445454646474748484949505051515252535354545555565657575858595960606161626263636464656566666767686869697070717172727373747475757676777778787979808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "End time: Wed Oct 23 15:31:18 2024\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[107], line 24\u001b[0m\n\u001b[1;32m     21\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mDropping SNP: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mrs_id\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     22\u001b[0m num \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mint\u001b[39m(rs_id\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m:\u001b[39m\u001b[38;5;124m'\u001b[39m)[\u001b[38;5;241m1\u001b[39m])\n\u001b[0;32m---> 24\u001b[0m \u001b[43mprocess_plink_data_with_drop\u001b[49m\u001b[43m(\u001b[49m\u001b[43mplink_file_test_prefix\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrs_id\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     26\u001b[0m result \u001b[38;5;241m=\u001b[39m subprocess\u001b[38;5;241m.\u001b[39mrun([\n\u001b[1;32m     27\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m../impute5_v1.2.0/impute5_v1.2.0_static\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m     28\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m--h\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mplink_file_train_prefix\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m_xcf.bcf\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     34\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m--threads\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m32\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m     35\u001b[0m ])\n\u001b[1;32m     37\u001b[0m imputed_genotype_array \u001b[38;5;241m=\u001b[39m extract_imputed_genotype_array(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mimputed_\u001b[39m\u001b[38;5;132;01m{\u001b[39;00midx\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.vcf\u001b[39m\u001b[38;5;124m'\u001b[39m, rs_id)\n",
      "Cell \u001b[0;32mIn[106], line 79\u001b[0m, in \u001b[0;36mprocess_plink_data_with_drop\u001b[0;34m(plink_file_prefix, rs_ids)\u001b[0m\n\u001b[1;32m     74\u001b[0m subprocess\u001b[38;5;241m.\u001b[39mrun(bcftools_command)\n\u001b[1;32m     76\u001b[0m \u001b[38;5;66;03m# print(\"BCF conversion complete.\")\u001b[39;00m\n\u001b[1;32m     77\u001b[0m \n\u001b[1;32m     78\u001b[0m \u001b[38;5;66;03m# print(\"Adding AC field..\")\u001b[39;00m\n\u001b[0;32m---> 79\u001b[0m \u001b[43msubprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\n\u001b[1;32m     80\u001b[0m \u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m../bcftools/bcftools\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m+fill-tags\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mmodified_test.bcf\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m-Ob\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m-o\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mmodified_test_AC.bcf\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m--\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m-t\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mAN,AC\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\n\u001b[1;32m     81\u001b[0m \u001b[43m\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstdout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msubprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mDEVNULL\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     82\u001b[0m \u001b[38;5;66;03m# print(\"Adding index file..\")\u001b[39;00m\n\u001b[1;32m     83\u001b[0m subprocess\u001b[38;5;241m.\u001b[39mrun([\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m../bcftools/bcftools\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mindex\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmodified_test_AC.bcf\u001b[39m\u001b[38;5;124m'\u001b[39m])\n",
      "File \u001b[0;32m/scratch2/prateek/miniconda3/envs/pyjuice_env/lib/python3.9/subprocess.py:507\u001b[0m, in \u001b[0;36mrun\u001b[0;34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    505\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m Popen(\u001b[38;5;241m*\u001b[39mpopenargs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs) \u001b[38;5;28;01mas\u001b[39;00m process:\n\u001b[1;32m    506\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 507\u001b[0m         stdout, stderr \u001b[38;5;241m=\u001b[39m \u001b[43mprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcommunicate\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43minput\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    508\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m TimeoutExpired \u001b[38;5;28;01mas\u001b[39;00m exc:\n\u001b[1;32m    509\u001b[0m         process\u001b[38;5;241m.\u001b[39mkill()\n",
      "File \u001b[0;32m/scratch2/prateek/miniconda3/envs/pyjuice_env/lib/python3.9/subprocess.py:1126\u001b[0m, in \u001b[0;36mPopen.communicate\u001b[0;34m(self, input, timeout)\u001b[0m\n\u001b[1;32m   1124\u001b[0m         stderr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstderr\u001b[38;5;241m.\u001b[39mread()\n\u001b[1;32m   1125\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstderr\u001b[38;5;241m.\u001b[39mclose()\n\u001b[0;32m-> 1126\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1127\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1128\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m/scratch2/prateek/miniconda3/envs/pyjuice_env/lib/python3.9/subprocess.py:1189\u001b[0m, in \u001b[0;36mPopen.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1187\u001b[0m     endtime \u001b[38;5;241m=\u001b[39m _time() \u001b[38;5;241m+\u001b[39m timeout\n\u001b[1;32m   1188\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1189\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_wait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1190\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyboardInterrupt\u001b[39;00m:\n\u001b[1;32m   1191\u001b[0m     \u001b[38;5;66;03m# https://bugs.python.org/issue25942\u001b[39;00m\n\u001b[1;32m   1192\u001b[0m     \u001b[38;5;66;03m# The first keyboard interrupt waits briefly for the child to\u001b[39;00m\n\u001b[1;32m   1193\u001b[0m     \u001b[38;5;66;03m# exit under the common assumption that it also received the ^C\u001b[39;00m\n\u001b[1;32m   1194\u001b[0m     \u001b[38;5;66;03m# generated SIGINT and will exit rapidly.\u001b[39;00m\n\u001b[1;32m   1195\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m/scratch2/prateek/miniconda3/envs/pyjuice_env/lib/python3.9/subprocess.py:1933\u001b[0m, in \u001b[0;36mPopen._wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1931\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1932\u001b[0m     \u001b[38;5;28;01mbreak\u001b[39;00m  \u001b[38;5;66;03m# Another thread waited.\u001b[39;00m\n\u001b[0;32m-> 1933\u001b[0m (pid, sts) \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_try_wait\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1934\u001b[0m \u001b[38;5;66;03m# Check the pid and loop as waitpid has been known to\u001b[39;00m\n\u001b[1;32m   1935\u001b[0m \u001b[38;5;66;03m# return 0 even without WNOHANG in odd situations.\u001b[39;00m\n\u001b[1;32m   1936\u001b[0m \u001b[38;5;66;03m# http://bugs.python.org/issue14396.\u001b[39;00m\n\u001b[1;32m   1937\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m pid \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpid:\n",
      "File \u001b[0;32m/scratch2/prateek/miniconda3/envs/pyjuice_env/lib/python3.9/subprocess.py:1891\u001b[0m, in \u001b[0;36mPopen._try_wait\u001b[0;34m(self, wait_flags)\u001b[0m\n\u001b[1;32m   1889\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"All callers to this function MUST hold self._waitpid_lock.\"\"\"\u001b[39;00m\n\u001b[1;32m   1890\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1891\u001b[0m     (pid, sts) \u001b[38;5;241m=\u001b[39m \u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwaitpid\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpid\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mwait_flags\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1892\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mChildProcessError\u001b[39;00m:\n\u001b[1;32m   1893\u001b[0m     \u001b[38;5;66;03m# This happens if SIGCLD is set to be ignored or waiting\u001b[39;00m\n\u001b[1;32m   1894\u001b[0m     \u001b[38;5;66;03m# for child processes has otherwise been disabled for our\u001b[39;00m\n\u001b[1;32m   1895\u001b[0m     \u001b[38;5;66;03m# process.  This child is dead, we can't get the status.\u001b[39;00m\n\u001b[1;32m   1896\u001b[0m     pid \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpid\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "buffer_region = f\"6:{idx1}-{idx2}\"\n",
    "\n",
    "os.environ['BCFTOOLS_PLUGINS'] = '/scratch2/prateek/bcftools/plugins'\n",
    "\n",
    "plink_file_train_prefix = \"fourier_ls-chr6-1167_train\"\n",
    "plink_file_test_prefix = \"fourier_ls-chr6-1167_test\"\n",
    "\n",
    "process_plink_data(plink_file_train_prefix)\n",
    "\n",
    "bim_file = f\"{plink_file_test_prefix}.bim\"\n",
    "rs_ids = []\n",
    "\n",
    "with open(bim_file, 'r') as f:\n",
    "    for line in f:\n",
    "        fields = line.split()\n",
    "        rs_id = fields[1]\n",
    "        rs_ids.append(rs_id)\n",
    "\n",
    "r2s = []\n",
    "for idx, rs_id in enumerate(rs_ids):\n",
    "    print(f\"Dropping SNP: {rs_id}\")\n",
    "    num = int(rs_id.split(':')[1])\n",
    "\n",
    "    process_plink_data_with_drop(plink_file_test_prefix, rs_id)\n",
    "    \n",
    "    result = subprocess.run([\n",
    "        '../impute5_v1.2.0/impute5_v1.2.0_static',\n",
    "        '--h', f'data/{plink_file_train_prefix}_xcf.bcf',\n",
    "        '--m', '../b37_recombination_maps/chr6.b37.gmap.gz',\n",
    "        '--g', 'modified_test_xcf.bcf',\n",
    "        '--r', f\"6:{num}-{num+1}\",\n",
    "        '--buffer-region', buffer_region,\n",
    "        '--o', f\"imputed_{idx}.vcf\",\n",
    "        '--l', f\"imputed_{idx}.log\",\n",
    "        '--threads', '32'\n",
    "    ])\n",
    "    \n",
    "    imputed_genotype_array = extract_imputed_genotype_array(f'imputed_{idx}.vcf', rs_id)\n",
    "    test_genotype_array = extract_test_genotype_array('data/fourier_ls-chr6-1167_test.vcf', rs_id)\n",
    "\n",
    "    r2 = compute_r_squared(imputed_genotype_array, test_genotype_array)\n",
    "    print(r2)\n",
    "    r2s.append(r2)\n",
    "    \n",
    "    # Delete the generated files after the SNP is processed\n",
    "    files_to_delete = [f\"imputed_{idx}.log\", f\"imputed_{idx}.vcf\", \"modified_test_AC.bcf\", \"modified_test_xcf.vcf\", \"modified_test_xcf.bcf.csi\", \"modified_test_xcf.bin\", \"modified_test_xcf.fam\", \"modified_test.bcf\", \"modified_test.vcf\"]\n",
    "    for file in files_to_delete:\n",
    "        if os.path.exists(file):\n",
    "            os.remove(file)\n",
    "        else:\n",
    "            print(f\"File {file} not found for deletion\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.4888814999377982]\n"
     ]
    }
   ],
   "source": [
    "print(r2s)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyjuice_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
